### PR TITLE
UI: Refactor Virtual Camera source selector dialog

### PR DESCRIPTION
**Pull request by tytan652**
### Description

Prepare the dialog to be extended with new feature like #5383 by moving some code elsewhere.

Most of the "dialog" code does not belong there and was moved.
- Struct & enum moved to their own header.
- The config code moved to the main window code.
- The video mix code is moved to output handler.

Also a fix the issue where if a scene/source is renamed, it is not in the vcam config.

### Motivation and Context

Avoid putting those change directly in #5383 and allow a separate review.

### How Has This Been Tested?

I tested:
- If the config was saved
- Starting the VCam with a source/scene/internal
- Changing the config while the VCam is active.
- Re-starting the VCam

### Types of changes

- Tweak (non-breaking change to improve existing functionality)

- Code cleanup (non-breaking change which makes code smaller or more readable)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by tytan652** [comment]
Rebase with refactor of #7374.

**Comment by tytan652** [comment]
Added a fix when a set scene/source is renamed.

**Comment by tytan652** [comment]
Just made a rebase and added a switch to avoid saving uneeded value.

File: UI/window-basic-main-outputs.cpp [hunk 1]
```diff
@@ -263,6 +271,85 @@ bool BasicOutputHandler::VirtualCamActive() const
...
+			obs_source_get_ref(s);
+			source = s.Get();
+			break;
+		}
+		break;
+	case VCamOutputType::SceneOutput:
+		source = obs_get_source_by_name(main->vcamConfig.scene.c_str());
+		break;
+	case VCamOutputType::SourceOutput:
+		OBSSource s =
```
Comment (tt2468) [inline UI/window-basic-main-outputs.cpp:298]: `obs_get_source_by_name()` returns a new reference already. `OBSSourceAutoRelease` should be used, and the `obs_source_release()` call below removed.

**Review by tt2468 — CHANGES_REQUESTED** [review]

File: UI/window-basic-main-outputs.cpp (hunk 1 above)
Comment (tytan652) [reply UI/window-basic-main-outputs.cpp:298]: It was like that in the original code, so this review should have been done in the original PR.

I just mostly moved the code for this part. I am thinking of considering this out of scope of this PR.

File: UI/window-basic-main-outputs.cpp [hunk 2]
```diff
@@ -229,30 +226,41 @@ inline BasicOutputHandler::BasicOutputHandler(OBSBasic *main_) : main(main_)
...
-		bool success = obs_output_start(virtualCam);
+	UpdateVirtualCamOutputSource();
 
-		if (!success)
-			OBSBasicVCamConfig::StopVideo();
+	if (!virtualCamVideo) {
+		virtualCamVideo = obs_view_add(virtualCamView);
 
-		return success;
+		if (!virtualCamVideo)
```
[the remark this answers is not in the document]
Comment (tytan652) [reply UI/window-basic-main-outputs.cpp:240]: It looks like it was the same with the original PR.

File: UI/window-basic-main-outputs.cpp [hunk 3]
```diff
@@ -263,6 +271,85 @@ bool BasicOutputHandler::VirtualCamActive() const
...
+	}
+
+	OBSSourceAutoRelease current = obs_view_get_source(virtualCamView, 0);
+	if (source != current)
+		obs_view_set_source(virtualCamView, 0, source);
+}
+
+void BasicOutputHandler::DestroyVirtualCamView()
+{
+	obs_view_remove(virtualCamView);
```
[the remark this answers is not in the document]
Comment (tytan652) [reply UI/window-basic-main-outputs.cpp:337]: It's normally almost the same as the original PR.

**Comment by tytan652** [comment]
I fixed some mistakes found by @tt2468.

But it looks other findings were already in #6577. I'll take my time to think about whether this PR should just move the code and fix some obvious problem or go beyond the scope and fix things that should have been find earlier. on the original.

**Comment by tytan652** [comment]
I thought about and I decided to not change the scope of PR.

So those conversation are considered out of scope:
- https://github.com/obsproject/obs-studio/pull/7169#discussion_r1071061835
- https://github.com/obsproject/obs-studio/pull/7169#discussion_r1071073805
- https://github.com/obsproject/obs-studio/pull/7169#discussion_r1071110040

**Comment by tt2468** [comment]
Current testing results:

No more freezing virtualcam output, everything largely looks good except for one issue I noticed. I do not know if this is an existing memory leak bug or a result of this PR but I'll describe the reproduction steps:

- Open OBS
- Create a new blank scene (yes, this is required)
- Create any source like `Color Source` in that scene
- Start virtualcam output and select `Color Source` as its input
- Delete the scene you created
  - The source is not released from the output, this is not inherently an issue though, just a note
- Open the virtualcam source selection dialog again and select a new source
- If you attempt to create another `Color Source`, you will see that the original one still exists, even though it is not added as a scene item anywhere.

**Comment by tytan652** [comment]
> * Create a new blank scene (yes, this is required)
> 
> * Create any source like `Color Source` in that scene
> 
> * Switch virtualcam to using `Color Source` as its input
> 
> * Delete the scene you created with the `Color Source` in it
>   
>   * The source is not released from the output, this is not inherently an issue though, just a note
> 
> * Open the virtualcam source selection dialog again and select a new source
> 
> * If you attempt to create another `Color Source`, you will see that the original one still exists, even though it is not added as a scene item anywhere.

I didn't dive into it yet, but it really looks like the source retention bug that is present since the original version of the feature.

But now I have more knowledge about OBS Signals so I have an idea on changing the source if deleted.

**Review by tt2468 — APPROVED** [review]

**Pull request merged**